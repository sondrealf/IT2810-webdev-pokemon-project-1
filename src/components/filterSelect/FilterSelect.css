.filter-select {
  position: relative;
  display: inline-block;
  width: 100%;
  height: 3em;
  max-width: 300px;
  background-color: #fff;
  color: #000;
  border: 0;
  border-radius: 5px;
  box-shadow: 0 0 0 1px #e5e5e5;
  padding: 0 1em;
  font-size: 16px;
}
