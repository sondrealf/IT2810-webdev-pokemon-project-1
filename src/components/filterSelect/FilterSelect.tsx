import { ChangeEvent } from "react";
import "./FilterSelect.css";

/**
 * Capitalizes the first letter of a string.
 * @param str - The string to capitalize.
 * @returns The capitalized string.
 */
const Capitalize = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

/**
 * A component that renders a select element for filtering options.
 * @param options - An array of strings representing the available options.
 * @param selected - A string representing the currently selected option.
 * @param handleChange - A function that handles the change event of the select element.
 * @param label - A string representing the label of the select element.
 * @returns {JSX.Element} - A React component that renders a select element for filtering options.
 */
const FilterSelect = ({
  options,
  selected,
  handleChange,
  label,
}: {
  options: string[];
  selected: string;
  handleChange: (e: ChangeEvent<HTMLSelectElement>) => void;
  label: string;
}) => {
  return (
    <div className="app__filtering-container">
      <span className="app__filtering-text">{label}</span>
      <select
        value={selected}
        onChange={handleChange}
        className="filter-select"
        placeholder="Select a color"
      >
        {options.map((option, index) => {
          if (option === "none") {
            return (
              <option key={`${option}-${index}`} value={option}>
                Select a color
              </option>
            );
          }
          return (
            <option key={`${option}-${index}`} value={option}>
              {Capitalize(option)}
            </option>
          );
        })}
      </select>
    </div>
  );
};

export default FilterSelect;
