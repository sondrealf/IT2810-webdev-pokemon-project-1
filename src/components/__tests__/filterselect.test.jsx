import { render, screen } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import FilterSelect from "../filterSelect/FilterSelect";
import { fireEvent } from "@testing-library/dom";

const mockTest = vi.fn();

describe("FilterSelect", () => {
  test("should match snapshot", () => {
    const { container } = render(
      <FilterSelect
        options={["test", "test2"]}
        handleChange={(e) => {
          console.log(e.target.value);
        }}
        selected="test"
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should render correctly", () => {
    render(
      <FilterSelect
        options={["none", "test", "test2"]}
        handleChange={(e) => {
          console.log(e.target.value);
        }}
        selected="test"
      />
    );

    expect(screen.getByText("Test")).toBeInTheDocument();
    expect(screen.getByText("Select a color")).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toHaveValue("test");
    expect(screen.getAllByRole("option")).toHaveLength(3);
  });

  test("should change option", () => {
    render(
      <FilterSelect
        options={["test", "test2"]}
        handleChange={(e) => mockTest(e.target.value)}
        selected="test"
      />
    );

    const select = screen.getByRole("combobox");

    fireEvent.change(select, { target: { value: "test2" } });
    expect(mockTest).toHaveBeenCalledWith("test2");
    expect(mockTest).toHaveBeenCalled();
  });
});
