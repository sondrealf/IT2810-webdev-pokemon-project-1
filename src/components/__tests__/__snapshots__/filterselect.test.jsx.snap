// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterSelect > should match snapshot 1`] = `
<div>
  <div
    class="app__filtering-container"
  >
    <span
      class="app__filtering-text"
    />
    <select
      class="filter-select"
      placeholder="Select a color"
    >
      <option
        value="test"
      >
        Test
      </option>
      <option
        value="test2"
      >
        Test2
      </option>
    </select>
  </div>
</div>
`;
